_posts/2020-05-11-microservices-on-axon.md

```diff
@@ -0,0 +1,198 @@
+Axon bierze na barki zarządzanie przepływem wszystkich informacji między komponentami np. kierowanie commandów do odpowiednich agregatów, czy zapisywanie eventów w event store. 
+Jeżeli chodzi o kwestie event store'a to framework zostawia tu pełną dowolność, choć nie każda baza spełni się w tej roli.
+Dodatkowym plusem jest bezproblemowa integracja ze Spring Bootem, możliwość skalowania i gotowość produkcyjna co moim zdaniem plasuje Axona jako mocnego gracza.
+
+# Event store
+Fundamentem projektu opartego o Event Sourcing jest oczywiście event store - źródło prawdy całego systemu, stąd wybór narzędzia pod tą funkcję jest kluczowy.
+
+### Może Kafka?
+Kafka opiera się na eventach, których kolejność pojawiania się może zostać zachowana - co zapobiega sytuacji, w której wykonamy aktualizację krotki, zanim zostanie ona utworzona.
+Ponadto Kafka trzyma dane na topicach, zapamiętując offset (liczbę porządkową) dla każdego eventu. Znając offset istnieje możliwość odtworzenia topica od tego offsetu, aż do końca - umożliwia to odtwarzanie idealnego stanu aplikacji dosłownie na zawołanie (dopóki nie mamy do czynienia z paruset milionami eventów :wink:).
+Do tego Kafka bardzo łatwo się skaluje oraz uniemożliwia edycję nałożonych eventów, co niewątpliwie jest plusem. Jest jednak parę punktów, które brakują Kafce, do bycia idealnym kandydatem na event store:
+- Problem pojawia się w momencie, gdy chcielibyśmy odtworzyć agregat na podstawie eventów. 
+Kafka w tym momencie musiałaby przeiterować cały topic od pewnego offsetu, aż do końca.
+W kolejnym kroku trzeba by odfiltrować te eventy, które nie dotyczą agregatu, który próbujemy odtworzyć, co wymaga od nas dodatkowej logiki w kodzie, oraz nakłada niepotrzebny dodatkowy narzut na event store (odfiltrowane eventy nie są nam potrzebne).
+- Drugim problemem jest brak natywnego wsparcia dla mechanizmu snapshotów, bez którego odtwarzanie stanu przy dużym narzucie zdarzeń może trwać wieki.
+
+Potencjalnym rozwiązaniem pierwszego problemu mógłby być osobny topic dla każdego agregatu, wówczas odpada konieczność filtrowania eventów.
+To rozwiązanie jednak może nie sprawdzić się przy ogromnej ilości agregatów. 
+Wynika to ze sposobu, w jaki Kafka przechowuje topici (a właściwie partycje) - dla każdej tworzony jest osobny katalog w systemie plików. 
+Szczegółowe wyjaśnienie znajduje się w [**filmie**](https://youtu.be/zUSWsJteRfw?t=2179) przygotowanym przez AxonIQ (firma odpowiedzialna za rozwój Axona).
+
+### AxonServer
+W kwestii event store AxonIQ wyszedł na przeciw potrzebom dając do dyspozycji swoje narzędzie, które idealnie spełnia się w tej roli - AxonServer:
+- pozwala na dokładanie eventów (z jednoczesnym brakiem możliwości edycji już istniejących)
+- zapewnia stałą wydajność niezależnie od ilości danych przetrzymywanych w event store
+- umożliwia konstruowanie snapshotów dla agregatów i nakładanie ich (w przypadku dużej ilości eventów rekonstrukcja agregatu bez funkcjonalności snapshotów może trochę trwać)
```

Comment:
Wymienianie powinno kończyć się każdorazowo jakimś znakiem interpunkcyjnym- np przecinkiem, a jeżeli występują przecinki w tekście to średnikiem lub kropką. Wcześniej w tekście użyłeś opcji z kropką, więc rekomenduję trzymanie się tej formy.